internal/lsp/server.go

```diff
@@ -204,40 +264,45 @@ func (s *Server) Run() error {
 
 func (s *Server) readLoop(ctx context.Context) error {
 	for {
-		msg, err := s.read()
-		if err != nil {
-			if errors.Is(err, lsproto.ErrInvalidRequest) {
-				s.sendError(nil, err)
-				continue
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
```

Comment:
For this, you don't need a select; just do `if err := ctx.Err(); err != nil { return er }`.